genesis/engine/entities/rigid_entity/rigid_entity.py

```diff
@@ -1620,6 +1640,128 @@ def plan_path(
     # ------------------------------------------------------------------------------------
+    def process_input(self, in_backward=False):
+        if in_backward:
+            # use negative index because buffer length might not be full
+            index = self._sim.cur_step_local - self._sim._steps_local
+            self._tgt = self._tgt_buffer[index].copy()
+        else:
+            self._tgt_buffer.append(self._tgt.copy())
+
+        # Apply targets sequentially
+        _tgt = self._tgt.copy()
+        for tgt in _tgt:
+            k = tgt["key"]
+            assert k in self._tgt_keys, f"Invalid target key: {k} not in {self._tgt_keys}"
+
+            # We do not need zero velocity here because if it was true, [set_dofs_velocity] from zero_velocity would
+            # be in [tgt]
+            zero_velocity = False
+            if k == "pos":
+                _pos = tgt["pos"]
+                _envs_idx = tgt["envs_idx"]
+                _relative = tgt["relative"]
+                _unsafe = tgt["unsafe"]
+
+                self.set_pos(_pos, envs_idx=_envs_idx, relative=_relative, zero_velocity=zero_velocity, unsafe=_unsafe)
+            elif k == "quat":
+                _quat = tgt["quat"]
+                _envs_idx = tgt["envs_idx"]
+                _relative = tgt["relative"]
+                _unsafe = tgt["unsafe"]
+
+                self.set_quat(
+                    _quat, envs_idx=_envs_idx, relative=_relative, zero_velocity=zero_velocity, unsafe=_unsafe
+                )
+            elif k == "qpos":
+                _qpos = tgt["qpos"]
+                _qs_idx_local = tgt["qs_idx_local"]
+                _envs_idx = tgt["envs_idx"]
+                _unsafe = tgt["unsafe"]
+
+                self.set_qpos(
+                    _qpos, qs_idx_local=_qs_idx_local, envs_idx=_envs_idx, zero_velocity=zero_velocity, unsafe=_unsafe
+                )
+            elif k == "dofs_velocity":
+                _velocity = tgt["velocity"]
+                _dofs_idx_local = tgt["dofs_idx_local"]
+                _envs_idx = tgt["envs_idx"]
+                _unsafe = tgt["unsafe"]
+
+                self.set_dofs_velocity(_velocity, dofs_idx_local=_dofs_idx_local, envs_idx=_envs_idx, unsafe=_unsafe)
+
+        self._tgt = []
+
+    def process_input_grad(self):
+        index = self._sim.cur_step_local - self._sim._steps_local
+        _tgt = self._tgt_buffer[index].copy()
+
+        for tgt in reversed(_tgt):
+            k = tgt["key"]
+            assert k in self._tgt_keys, f"Invalid target key: {k} not in {self._tgt_keys}"
+            if k == "pos":
+                _pos = tgt["pos"]
+                _envs_idx = tgt["envs_idx"]
+                _relative = tgt["relative"]
+                _unsafe = tgt["unsafe"]
+
+                if _pos is not None and _pos.requires_grad:
+                    _pos._backward_from_ti(self.set_pos_grad, _envs_idx, _relative, _unsafe)
```

Comment:
Why are you unpacking all these arguments to pass them back? It would be more clear and future proof to just do:
```python
_pos._backward_from_ti(self.set_pos_grad, **tgt)
```